utopia-vscode-extension/src/extension.ts

```diff
@@ -219,20 +225,36 @@ function cursorPositionChanged(event: vscode.TextEditorSelectionChangeEvent): vo
   sendMessage(editorCursorPositionChanged(filename, position.line, position.character))
 }
 
-function revealRangeIfPossible(boundsInFile: BoundsInFile): void {
-  const visibleEditors = vscode.window.visibleTextEditors
-  for (const visibleEditor of visibleEditors) {
-    const filename = visibleEditor.document.uri.path
-    if (boundsInFile.filePath === filename) {
-      const rangeToReveal = getVSCodeRange(boundsInFile)
-      const alreadySelected = rangeToReveal.contains(visibleEditor.selection)
-      const alreadyVisible = visibleEditor.visibleRanges.some((r) =>
-        r.contains(visibleEditor.selection),
+async function revealRangeIfPossible(
+  workspaceRootUri: vscode.Uri,
+  boundsInFile: BoundsInFile,
+): Promise<void> {
+  const visibleEditor = vscode.window.visibleTextEditors.find(
+    (editor) => editor.document.uri.path === boundsInFile.filePath,
+  )
+  if (visibleEditor == null) {
+    const opened = await openFile(vscode.Uri.joinPath(workspaceRootUri, boundsInFile.filePath))
+    if (opened) {
+      revealRangeIfPossible(workspaceRootUri, boundsInFile)
+    }
+  } else {
+    const rangeToReveal = getVSCodeRangeForScrolling(boundsInFile)
+    const alreadySelected = rangeToReveal.contains(visibleEditor.selection)
+    const alreadyVisible = visibleEditor.visibleRanges.some((r) =>
+      r.contains(visibleEditor.selection),
+    )
+
+    if (!alreadySelected) {
+      const selectionRange = getVSCodeRange(boundsInFile)
+      visibleEditor.selection = new vscode.Selection(selectionRange.start, selectionRange.start)
+    }
+
+    const shouldReveal = !(alreadySelected && alreadyVisible)
+    if (shouldReveal) {
+      visibleEditor.revealRange(
+        rangeToReveal,
+        vscode.TextEditorRevealType.InCenterIfOutsideViewport,
```

Comment:
I changed the reveal type here from the default, as this felt more natural to me when playing with this. The default behaviour is to reveal the code with minimal scrolling, which either opens it at the very top or the very bottom. This behaviour will reveal the code in the center (but only if the code was previously not visible). I don't mind changing this back though if others dislike it.